### PR TITLE
Detection in deblending_step function and no more npeaks_per_iteration

### PR DESCRIPTION
**Pull request by BastienArcelin**
Answering #9 

I also added two files in the data folder for the center of the two example fields that are used at the end of the notebook and cleaned the notebook. I added the comparison of what is obtained with the deblending procedure compared to with a deblending with detection with SExtractor.

### CODE REVIEW

File: debvader/debvader.py [hunk 1]
```diff
@@ -8,12 +8,15 @@
...
 import sep
 
-from debvader import model
-
+#from debvader import model
+###### TO SUPRESSS AND UNCOMMENT PREVIOUS LINES
+import sys
+sys.path.insert(0,'.')
+import model
+######
```
Comment (aboucaud) [inline debvader/debvader.py:19]: Whaaaat ?!

File: debvader/debvader.py [hunk 2]
```diff
@@ -157,28 +161,30 @@ def fun (x, img, net_output):
                 denoised_field_std[:,:,j] +=scipy.ndimage.shift(output_images_stddev_padded[i,:,:,j],shift = (galaxy_distances_to_center[k][0]+opt.x[0],galaxy_distances_to_center[k][1]+opt.x[1]))
                 if epistemic_uncertainty_estimation:
                     denoised_field_epistemic[:,:,j] +=scipy.ndimage.shift(output_images_epistemic_padded[i,:,:,j],shift = (galaxy_distances_to_center[k][0]+opt.x[0],galaxy_distances_to_center[k][1]+opt.x[1]))
-                if ((np.sum(output_images_epistemic_padded[i,:,:,2])/np.sum(output_images_mean_padded[i,:,:,2]))>epistemic_criterion) or (metrics.mean_squared_error(cutout_images[k],output_images_mean[i])>mse_criterion): # avoid to add galaxies generated with too high uncertainty
+                if ((np.sum(output_images_epistemic_padded[i,:,:,2])/np.sum(output_images_mean_padded[i,:,:,2]))>epistemic_criterion) or (metrics.mean_squared_error(cutout_images[k,int(cutout_size/2)-5:int(cutout_size/2)+5,int(cutout_size/2)-5:int(cutout_size/2)+5,:],output_images_mean[i,int(cutout_size/2)-5:int(cutout_size/2)+5,int(cutout_size/2)-5:int(cutout_size/2)+5,:])>mse_criterion): # avoid to add galaxies generated with too high uncertainty
```
Comment (aboucaud) [inline debvader/debvader.py:164]: intermediate calculus would be welcome

File: debvader/debvader.py [hunk 3]
```diff
@@ -110,7 +113,7 @@ def deblend_field(net, field_image, galaxy_distances_to_center, cutout_images =
         output_images_mean, output_images_distribution = deblend(net, cutout_images[list_idx], normalised=normalised)  
     if list_idx==[]:
         print('No galaxy deblended. End of the iterative procedure.')
-        return field_image, field_image, np.zeros((field_size,field_size,nb_of_bands)), np.zeros((field_size,field_size,nb_of_bands)), np.zeros((field_size,field_size,nb_of_bands)), cutout_images, output_images_mean, output_images_distribution, np.zeros((len(output_images_mean),2)), list_idx
+        return field_image, field_image, np.zeros((field_size,field_size,nb_of_bands)), np.zeros((field_size,field_size,nb_of_bands)), np.zeros((field_size,field_size,nb_of_bands)), cutout_images, output_images_mean, output_images_distribution, np.zeros((len(output_images_mean),2)), list_idx, 0
```
Comment (aboucaud) [inline debvader/debvader.py:116]: maybe the return value should be simplified.
dict ?!
dataclass ?!

File: debvader/debvader.py [hunk 4]
```diff
@@ -157,28 +161,30 @@ def fun (x, img, net_output):
...
+                if ((np.sum(output_images_epistemic_padded[i,:,:,2])/np.sum(output_images_mean_padded[i,:,:,2]))>epistemic_criterion) or (metrics.mean_squared_error(cutout_images[k,int(cutout_size/2)-5:int(cutout_size/2)+5,int(cutout_size/2)-5:int(cutout_size/2)+5,:],output_images_mean[i,int(cutout_size/2)-5:int(cutout_size/2)+5,int(cutout_size/2)-5:int(cutout_size/2)+5,:])>mse_criterion): # avoid to add galaxies generated with too high uncertainty
                     pass
                 else:
                     field_image[0,:,:,j] -= scipy.ndimage.shift(output_images_mean_padded[i,:,:,j],shift = (galaxy_distances_to_center[k][0],galaxy_distances_to_center[k][1]))
                     denoised_field[:,:,j] +=scipy.ndimage.shift(output_images_mean_padded[i,:,:,j],shift = (galaxy_distances_to_center[k][0],galaxy_distances_to_center[k][1]))
+                    nb_of_galaxies_in_deblended_field+=1/nb_of_bands
```
Comment (aboucaud) [inline debvader/debvader.py:180]: `nb_of_..` usually refers to an integer but here the increment is a float..

File: debvader/debvader.py [hunk 5]
```diff
@@ -209,25 +210,15 @@ def iterative_deblending(net, field_image, galaxy_distances_to_center_in, npeaks
...
 
-        field_img_save, field_image, denoised_field, denoised_field_std, denoised_field_epistemic, cutout_images, output_images_mean, output_images_distribution, shifts, galaxy_distances_to_center, mse_step = deblending_step(net, field_image, detection_k, cutout_images = None, cutout_size = cutout_size, nb_of_bands = nb_of_bands, optimise_positions=optimise_positions, epistemic_uncertainty_estimation=epistemic_uncertainty_estimation, epistemic_criterion=epistemic_criterion, mse_criterion=mse_criterion, normalised=normalised)
+        field_img_save, field_image, denoised_field, denoised_field_std, denoised_field_epistemic, cutout_images, output_images_mean, output_images_distribution, shifts, galaxy_distances_to_center_total, mse_step, nb_of_galaxies_in_deblended_field = deblending_step(net, field_image, galaxy_distances_to_center_total, cutout_images = None, cutout_size = cutout_size, nb_of_bands = nb_of_bands, optimise_positions=optimise_positions, epistemic_uncertainty_estimation=epistemic_uncertainty_estimation, epistemic_criterion=epistemic_criterion, mse_criterion=mse_criterion, normalised=normalised,factor=factor)
```
Comment (aboucaud) [inline debvader/debvader.py:221]: here again, way too complicated, and too easy to make a mistake

File: debvader/debvader.py [hunk 6]
```diff
@@ -238,32 +229,19 @@ def iterative_deblending(net, field_image, galaxy_distances_to_center_in, npeaks
...
-                if detection_k[i] in galaxy_distances_to_center_total:
-                    idx_to_remove.append(i)
-            detection_k = np.delete(detection_k, idx_to_remove, axis = 0)
-            if npeaks_per_iteration==500:
-                print('converged on maximum peak per iteration.')
-                break
-
-        print(str(len(shifts))+' galaxies found up to this step.')
+
+        print(str(nb_of_galaxies_in_deblended_field_total)+' galaxies found up to this step.')
```
Comment (aboucaud) [inline debvader/debvader.py:236]:
```suggestion
        print(f'{nb_of_galaxies_in_deblended_field_total} galaxies found up to this step.')
```

**Review by aboucaud — COMMENTED** [review]
The return values of the main functions should be easily explored through an interactive session.
Therefore any value should be accessible from its name or as an attribute : 
`result['field_image']` or `result.field_image`

File: debvader/debvader.py (hunk 3 above)
Comment (b-biswas) [reply debvader/debvader.py:116]: Maybe we should also start adding the return variables in the docstring?

File: debvader/debvader.py (hunk 2 above)
Comment (BastienArcelin) [reply debvader/debvader.py]: Ok changed in modified in commit [b085507](https://github.com/BastienArcelin/debvader/commit/b085507933267ac5fdcff4a486cde70d6985bc94#diff-a5dae2bb532104d1ce10a8b52d5c651b023b7d0d57e87a12c0c6942e0d44cee7)

File: debvader/debvader.py (hunk 1 above)
Comment (BastienArcelin) [reply debvader/debvader.py]: Old version. Modified in commit [b085507](https://github.com/BastienArcelin/debvader/commit/b085507933267ac5fdcff4a486cde70d6985bc94#diff-a5dae2bb532104d1ce10a8b52d5c651b023b7d0d57e87a12c0c6942e0d44cee7)

File: debvader/debvader.py (hunk 4 above)
Comment (BastienArcelin) [reply debvader/debvader.py]: Right. I changed it in commit [7fc7b7a](https://github.com/BastienArcelin/debvader/commit/7fc7b7a241d09535529bca1d564469a252fe0fe8#diff-a5dae2bb532104d1ce10a8b52d5c651b023b7d0d57e87a12c0c6942e0d44cee7)

File: debvader/debvader.py (hunk 3 above)
Comment (BastienArcelin) [reply debvader/debvader.py]: True. Changed in [7754ad3](https://github.com/BastienArcelin/debvader/commit/7754ad39feb606cfee8e4b0f776da5645eb16258#diff-a5dae2bb532104d1ce10a8b52d5c651b023b7d0d57e87a12c0c6942e0d44cee7) with corresponding changes in the notebook

File: debvader/debvader.py (hunk 5 above)
Comment (BastienArcelin) [reply debvader/debvader.py]: Same, see [7754ad3](https://github.com/BastienArcelin/debvader/commit/7754ad39feb606cfee8e4b0f776da5645eb16258#diff-a5dae2bb532104d1ce10a8b52d5c651b023b7d0d57e87a12c0c6942e0d44cee7)

File: debvader/debvader.py (hunk 6 above)
Comment (BastienArcelin) [reply debvader/debvader.py]: Ok I tried to use more f-strings. See [1387371](https://github.com/BastienArcelin/debvader/commit/138737139df0104610b68f98323b9dab51d393a1#diff-a5dae2bb532104d1ce10a8b52d5c651b023b7d0d57e87a12c0c6942e0d44cee7)

**Comment by BastienArcelin** [comment]
That is a good question actually. My guess is that this comes more from the fact that we are generating galaxies with a higher flux that we expect as we our deblender is not trained on faint galaxies. 
If I am not mistaken, the galaxies that are found by debvader in the second step are real galaxies, but some of them are very faint

**Review by aboucaud — APPROVED** [review]
Looks already much cleaner!

I have a few more general comments on the code but I guess this can be merged the way it is now. 

In general, keep in mind that a "clean code" is a code where the reader needs a minimum of effort to understand every single line. And I feel like there are many lines that I struggle to understand without spending some time browsing around and detailing parts of it.

**Pull request merged**